### ISSUE TITLE
[Feature] Service Discovery support local cache when registry is unavailable

### ISSUE THREAD

**Issue by AlbumenJ** [apache/dubbo#13856]
### Pre-check

- [X] I am sure that all the content I provide is in English.

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dubbo/issues?q=is%3Aissue) and found no similar feature requirement.

### Apache Dubbo Component

Java SDK (apache/dubbo)

### Descriptions

Service Discovery support local cache when registry is unavailable

### Are you willing to submit a pull request to fix on your own?

- [ ] Yes I am willing to submit a pull request on my own!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by finefuture** [comment]
What scenarios need this feature? In general, the registry sdk already supports local cache.

**Comment by AlbumenJ** [comment]
> What scenarios need this feature? In general, the registry sdk already supports local cache.

1. zookeeper not support
2. Nacos's local cache is not useful for Dubbo. Nacos will use local cache by highest priority, but Dubbo only want it as the fallback.

**Comment by FAWC438** [comment]
Hi, I am new to Dubbo community and willing to complete this task. 

Is there any restriction on this issue? For example, in terms of completion time and code quality.

**Comment by AlbumenJ** [comment]
> Hi, I am new to Dubbo community and willing to complete this task.
> 
> Is there any restriction on this issue? For example, in terms of completion time and code quality.

1. We have no completion time limitation
2. For code quality, we have add a lot of checkers in Github Actions and we will check it when pull request reviewing

**Comment by FAWC438** [comment]
Hi @AlbumenJ, I checked the relevant code and documentation, and I think the "local cache" mentioned in this issue actually refers to the consumer caching information such as the provider's `ip:port` obtained from the registry that is not down. Is that correct? 

I did see the local caching function of `metadata` in the code, but as far as I know from reading the documentation, the metadata link maintained by Dubbo does not actually require the participation of the Service Discovery registry(Nacos or Zookeeper etc.), so does the work of this issue actually have nothing to do with `metadata` ?

And, regarding the implementation of local cache, since you mentioned that it will be used as a fallback solution, I think it only needs to be implemented in memory (for example, through a `Map` ) without the need to store it in the disk (like Nacos) , because it is a small probability event after all.

**Comment by AlbumenJ** [comment]
> And, regarding the implementation of local cache, since you mentioned that it will be used as a fallback solution, I think it only needs to be implemented in memory (for example, through a `Map` ) without the need to store it in the disk (like Nacos) , because it is a small probability event after all.

The main purpose is to save in the disk. To prevent if registry crashed Dubbo cannot start.

**Comment by walkinggo** [comment]
> > And, regarding the implementation of local cache, since you mentioned that it will be used as a fallback solution, I think it only needs to be implemented in memory (for example, through a `Map` ) without the need to store it in the disk (like Nacos) , because it is a small probability event after all.
> 
> The main purpose is to save in the disk. To prevent if registry crashed Dubbo cannot start.

Indeed, you can use serialization interfaces to save information on the local disk. However, if users tamper with the files storing this information, it can lead to data security issues. Unauthorized changes might corrupt the data, introduce errors, or cause security vulnerabilities if sensitive data is exposed or manipulated.

**Comment by walkinggo** [comment]
Hey, I want to help too, can you provide more detailed information? For example, what is the local specific information storage location and storage form. @AlbumenJ

**Comment by AlbumenJ** [comment]
> Hey, I want to help too, can you provide more detailed information? For example, what is the local specific information storage location and storage form. @AlbumenJ

You can check the related example in `org.apache.dubbo.registry.support.AbstractRegistry`

**Comment by walkinggo** [comment]
Firstly,we should make the zookeeper support the local cache.Secondly,we should make the nacos implements the fallback.It's the all the tasks need to finish in this issue?

**Comment by AlbumenJ** [comment]
> Firstly,we should make the zookeeper support the local cache.Secondly,we should make the nacos implements the fallback.It's the all the tasks need to finish in this issue?

We can make it more common and not depend on the specific registry.

**Comment by walkinggo** [comment]
I test it, and i find out that dubbo cannot work if registry is down.I can check it is true when registry is zookeeper or nacos.And also,i can see the `org.apache.dubbo.registry.support.AbstractRegistry` will store the local cache when `localCacheEnabled `is true.I think the way to finish it is read the local cache when the registry cannot be connected.

**Comment by AlbumenJ** [comment]
> I test it, and i find out that dubbo cannot work if registry is down.I can check it is true when registry is zookeeper or nacos.And also,i can see the `org.apache.dubbo.registry.support.AbstractRegistry` will store the local cache when `localCacheEnabled `is true.I think the way to finish it is read the local cache when the registry cannot be connected.

Yep, you're right. BTW, we should make sure Dubbo will try to reconnect the broken registry continuously.

**Comment by HeyAlaia** [comment]
If Dubbo cannot reconnect to the broken registry, this may be another problem. should open another issue

**Comment by dyx1234** [comment]
I think I can solve this problem, please assign me～ @AlbumenJ

**Comment by AlbumenJ** [comment]
> I think I can solve this problem, please assign me～ @AlbumenJ

Please go ahead

**Comment by walkinggo** [comment]
Looks like this issue is still unresolved. Please assign it to me.i will go ahead @AlbumenJ

**Comment by AlbumenJ** [comment]
> Looks like this issue is still unresolved. Please assign it to me.i will go ahead @AlbumenJ

done

**Comment by MaxChen1123** [comment]
@AlbumenJ  Hi, is this issue still unresolved? Could you please assign it to me? I'd like to have a try.

**Comment by MaxChen1123** [comment]
@AlbumenJ  For `org.apache.dubbo.registry.zookeeper.ZookeeperRegistry`, if we want Dubbo can start even if the `zkClient` cannot connect to Zookeeper server,  we can just make the zkClient's initialization async (in this scenario, zkClient would be null ). So once the `doSubscribe` or `doRegister` is called, it will throw Exception, and the catch block in `org.apache.dubbo.registry.support.FailbackRegistry` will be executed, which will use the local cache in the disk. I've not tested the Nacos implement, but the solution may be the same.
But if we want to reconnect the registry continuously, do we need to retry all the failed call when finally connect to registry? Since FallbackRegistry already have the logic that will retry the failed call only once.

**Comment by chiangzeon** [comment]
@AlbumenJ Hi, is this issue still unresolved? Could you please assign it to me? I'd like to have a try.

**Comment by ghostly72** [comment]
Hi @AlbumenJ , i added some code in `org.apache.dubbo.registry.support.AbstractRegistry` to read from the local cache when registry is not available and notify the consumer abt it. 

submitted a pr as well
can u guys review it ? happy to get some suggestions on it 
thanks

**Comment by ShauryaChauhan1411** [comment]
Hi @AlbumenJ, I'm Shaurya. I recently merged #15872 (Nacos Registry fix). I'm very interested in Service Discovery and would like to help implement the local cache support for when the registry is unavailable. May I be assigned to this?

**Comment by ghostly72** [comment]
Hi @ShauryaChauhan1411 . I've linked a PR to this issue, and im working on it.

**Comment by ghostly72** [comment]
hi
right now, in the pr i created for this, my code goes like (inside abstractregistry.java)
```
if(localCacheEnabled && !isAvailable){
            try{
                List<URL> cached=getCacheUrls(url);
                if(CollectionUtils.isNotEmpty(cached)){
                    listener.notify(cached);
                    logger.info("Registry unavailable or not yet ready. Loaded cached URLs for "+url.getServiceKey()+" : "+cached.size());
                }
            }catch(Throwable t){
                logger.warn(INTERNAL_ERROR,"failed to load local cache","","Failed to load chached registry data for "+url,t);
                throw t;
            }
        }
```
but this fails tests because i think `isAvailable()` returns false not only when the registry crashes, but also when:
-the registry hasn’t connected yet (say during startup)
-the client object is null (like in tests or mocks)
-the registry was shut down gracefully

**Comment by ghostly72** [comment]
so to detect only the case of registry failure, can we add another flag like `isRegistryDown` in abstract registry?
im thinking of maybr adding:
```
private volatile boolean registryDown = false;

protected boolean isRegistryDown() {
    return registryDown;
}

protected void setRegistryDown(boolean down) {
    this.registryDown = down;
}
``` 
and 
```
private boolean isConnectionException(Throwable t) {
    return t instanceof java.io.IOException
        || t instanceof java.net.ConnectException
        || t instanceof java.net.SocketTimeoutException
        || (t.getMessage() != null && t.getMessage().contains("Connection refused"));
// we can add other exceptions that correspond to the registry failure or unavailability scenario as well
}
```
inside abstract registry. 
other registry implementatoins ultimately extend this.

so in every registry, we'll have code block similar to
```
try{
    // register/unregister/subscriber/unsubscribe to registry
}catch{
    // msg saying failed to connect due to some error
   // here we can call isConnectionException and check if we should set the flag or not accordingly
```

and then finally inside abstractregistry's code
we can have 
```
if(!isRegistryDown) enable local cache 
```

thoughts anyone?

**Comment by Anon2Tokyo** [comment]
Hello, is there anyone else who is working on solving this issue? I can handle this issue myself.
